Feat/max error rate

markVaykhansky opened the pull request:
Addressing this issue:
https://github.com/neuralmagic/guidellm/issues/105

`--max-error-rate` flag was added. 
It is only applicable in either or the the following scenarios:
1) Fixed length dataset
2) `max_requests` is set
2) `rate_type` is `constant` and `max_duration` is set

markurtz reviewed (changes requested):
I left a few minor comments for the less intrusive changes. Also, please be sure to run `tox -e style` and `tox -e types` locally and check those are passing -- there are a bunch of styling pieces, such as where parentheses are placed, how new lines are escaped, etc that are inconsistent with the ruff styling setup. Running tox -e style locally applied several fixes and raised some errors that couldn't be resolved without either modifying the code or ignoring them where appropriate.

For implementation, I think defaulting always to use the shutdown event is a good pattern rather than relying on two signals: either `None` being placed on the queue or the shutdown event being set if the maximum error rate is reached. I'd like to incorporate that into this to reduce complexity. 

Additionally, I think we should generally combine these into a single parameter, max_errors, which can be set as either max_error_rate or an absolute number of errors that can occur, allowing us to support all benchmarking pathways.

Additionally, with the above, as well as some combination logic in asyncio, we should be able to simplify the changes, so we don't need to make significant changes to the resolve requests pathways and can cancel for any type. For example, we should be able to replace the root process runner to call into async loops and handle the shutdown event globally in a central place with something like the following:
```python
    def run_process(
        self,
        type_: Literal["synchronous", "asynchronous"],
        requests_queue: multiprocessing.Queue,
        results_queue: multiprocessing.Queue,
        shutdown_event: multiprocessing.Event,
        shutdown_poll_interval: float,
        process_id: int,
        max_concurrency: int,
    ):
        async def _process_runner():
            if type_ == "synchronous":
                loop_task = asyncio.create_task(self.synchronous_loop(...))
            elif type_ == "asynchronous":
                loop_task = asyncio.create_task(self.asyncronous_loop(...))
            else:
                raise ValueError(f"Invalid process type: {type_}")
            
            shutdown_task = asyncio.create_task(
                self.wait_for_shutdown(shutdown_event, shutdown_poll_interval)
            )
            
            done, pending = await asyncio.wait(
                [
                    loop_task,
                    shutdown_task,
                ],
                return_when=asyncio.FIRST_EXCEPTION,
            )

            for task in pending:
                task.cancel()
                try:
                    await task
                except asyncio.CancelledError:
                    pass

            for task in done:
                if task.exception():
                    raise task.exception()  # need to check and ignore cancelations
        
        try:
            asyncio.run(_process_runner())
        except Exception as exc:  # noqa: BLE001
            logger.error(
                f"Error in worker process {process_id}: {exc}",
                exc_info=True,
                stack_info=True,
            )
        finally:
            shutdown_event.set()  # ensure shutdown event is set to stop other processes

    async def wait_for_shutdown(
        self, shutdown_event: multiprocessing.Event, shutdown_poll_interval: float
    ):
        while not shutdown_event.is_set():
            await asyncio.sleep(shutdown_poll_interval)

        raise asyncio.CancelledError("Shutdown event set, cancelling process loop.")
    
    async def synchronous_loop(self, ...):
        while True:
            process_request = await self.get_request(requests_queue)
            ...

    async def asyncronous_loop(self, ...):
        while True:
            process_request = await self.get_request(requests_queue)
            ...
```

markVaykhansky commented on src/guidellm/benchmark/benchmark.py:
```diff
@@ -213,6 +216,14 @@ class BenchmarkRunStats(StandardBaseModel):
+            "The number of errored requests divided by the number "
+            "of errored requests. This can be higher than max_error_rate "
+            "(if applicable) cause it does not take into "
+            "account incomplete requests."
```
Correct

markVaykhansky commented on src/guidellm/scheduler/scheduler.py:
```diff
@@ -159,6 +166,17 @@ async def run(
                         run_info,
                     )
                     if iter_result is not None:
+                        if iter_result.request_info.errored \
```
Done

markVaykhansky commented:
I've changed the logic as suggested to have a single termination method which is a `shutdown event`. 
However, this did not go a smoothly as expected:

After implementing the code like you suggested above I ran into an issue with `get_request`. After a shutdown was initiated and the `_wait_for_shutdown` task raised an exception the code moved to cancel the pending task i.e the request sending task. However, it was stuck indefinitely cause the logic there's `asyncio.to_thread(requests_queue.get)` which is blocking. To resolve the issue I've re-introduced the `sleep_intermittently` logic which check the `shutdown_event` periodically inside the `get_request`. Then I ran into another issue, after sending the shutdown event there's a race condition between the `asyncio.wait` and checking for errors in `get_request`, to resolve the issue I've created a separate `internal_shutdown_event` which is set after the `task.cancel()`. 

Another super weird thing that happened is the raising specifically `asyncio.CancellationError` in `_wait _for_shutdown` did not trigger the return of `asyncio.wait` with `FIRST_EXCEPTION` and it was stuck forever, I'm not sure why couldn't find a reasonable explanation. Ended up throwing a custom error instead. 

Also, I added the context of last X request as suggested but I feel that for the scenario where we do know the exact amount, i.e const rate + fixed duration, it is best to look at the entire run.
